File: packages/dart_node_core/lib/src/interop.dart
import 'dart:js_interop';
import 'dart:js_interop_unsafe';

extension type _GlobalContext(JSObject _) implements JSObject {
  external JSFunction get require;
  external JSObject get console;
}

extension type _Console(JSObject _) implements JSObject {
  external JSFunction get log;
}

_GlobalContext get _context => _GlobalContext(globalContext);

/// Get a value from the global JavaScript context
JSAny? getGlobal(String name) => globalContext[name];

/// Get Node's require function
JSFunction get require => _context.require;

/// Get the console object
JSObject get console => _context.console;

/// Log to console
void consoleLog(String message) {
  _Console(console).log.callAsFunction(null, message.toJS);
}

File: packages/dart_node_core/lib/src/interop.ts
// TypeScript version of interop.dart

// Global context interface
interface GlobalContext {
  require: Function;
  console: Console;
}

// Console interface
interface Console {
  log: Function;
  error: Function;
}

// Get the global context
const getGlobalContext = (): GlobalContext => {
  return global as unknown as GlobalContext;
};

// Get a value from the global JavaScript context
const getGlobal = (name: string): any => {
  return (global as any)[name];
};

// Get Node's require function
const require = getGlobalContext().require;

// Get the console object
const console = getGlobalContext().console;

// Log to console (stdout)
const consoleLog = (message: string): void => {
  console.log(message);
};

// Log to console.error (stderr)
const consoleError = (message: string): void => {
  console.error(message);
};

export {
  getGlobal,
  require,
  console,
  consoleLog,
  consoleError
};